### PR TITLE
Cbmc test

### LINKED ISSUE #413: Port CBMC unit tests to ESBMC
CBMC contains lots of unit tests using the `catch` framework. Most of those tests could be reused at ESBMC with minor changes. https://github.com/diffblue/cbmc/tree/develop/unit

Not only that would create a good base suite, it would be helpful to create [property-based](https://github.com/emil-e/rapidcheck) tests and fuzzing targets.

Also, although we already use Boost.Test I wouldn't mind changing it to catch2 since it is way simpler to understand.

### PR DESCRIPTION
**Pull request by rafaelsamenezes**
This PR:

- converts from Boost.Test into Catch2
- brings some unittests from cbmc that were easy to integrate
- Closes #413

### CODE REVIEW

**Review by mikhailramalho — APPROVED** [review]
I saw some TODOs here, are they yours or from CBMC? If the latter, I believe we can remove them.

Otherwise, LGTM.

**Review by lucasccordeiro — APPROVED** [review]

**Comment by rafaelsamenezes** [comment]
> I saw some TODOs here, are they yours or from CBMC? 

Mine I've added for methods that we may want to add in the future e.g. checking if a symbol is well formed

**Review by lucasccordeiro — APPROVED** [review]

File: unit/clang-c-frontend/typecast.fuzz.cpp [hunk 1]
```diff
@@ -21,7 +21,7 @@
 #include <clang-c-frontend/typecast.h>
 #include <util/type.h>
 #include <util/expr_util.h>
-#include "util.h"
+#include "../testing-utils/util_irep.h"
```
Comment (lucasccordeiro) [inline unit/clang-c-frontend/typecast.fuzz.cpp:24]: Is it correct?

File: unit/clang-c-frontend/typecast.test.cpp [hunk 2]
```diff
@@ -9,122 +9,92 @@
    - ToUnion (builtins, exceptions)
  \*******************************************************************/
 
-#define BOOST_TEST_MODULE "Typecast Test"
+#include "../testing-utils/util_irep.h"
```
Comment (lucasccordeiro) [inline unit/clang-c-frontend/typecast.test.cpp:12]: Is it correct?

File: unit/util/replace_symbol.test.cpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+  REQUIRE(!r.replace(binary));
+  REQUIRE(binary.op0() == other_expr);
+
+  REQUIRE(!r.replace(s1));
+  REQUIRE(s1 == other_expr);
+
+  REQUIRE(r.replace(s2));
+  REQUIRE(s2 == symbol_exprt("b", typet("some_type")));
+
+  // REQUIRE(!r.replace(array_type)); TODO: Fix this
```
Comment (lucasccordeiro) [inline unit/util/replace_symbol.test.cpp:40]: Shall these comments be removed?

File: unit/util/replace_symbol.test.cpp [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  // REQUIRE(array_type.size() == other_expr);
```
Comment (lucasccordeiro) [inline unit/util/replace_symbol.test.cpp:41]: Shall these comments be removed?

File: unit/util/replace_symbol.test.cpp [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    address_of_exprt(index),
+    typet("some_type"));
+
+  constant_exprt c("some_value", "some_value", typet("some_type"));
+
+  replace_symbolt r;
+  r.insert(s1.identifier(), c);
+
+  REQUIRE(!r.replace(binary));
+  // REQUIRE(binary.op0() == address_of_exprt(s1));
```
Comment (lucasccordeiro) [inline unit/util/replace_symbol.test.cpp:63]: Shall these comments be removed?

File: unit/util/replace_symbol.test.cpp [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+  const index_exprt &index_expr =
+    to_index_expr(to_address_of_expr(binary.op1()).object());
+  //REQUIRE(to_array_type(index_expr.array().type()).size() == c);
```
Comment (lucasccordeiro) [inline unit/util/replace_symbol.test.cpp:66]: Shall these comments be removed?

File: unit/util/replace_symbol.test.cpp [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  REQUIRE(index_expr.index() == c);
+
+  address_of_exprt address_of(s1);
+  // r.erase("a");
+  r.insert(s1.identifier(), address_of);
+
+  REQUIRE(r.replace(binary));
+  // REQUIRE(binary.op0() == address_of_exprt(s1));
```
Comment (lucasccordeiro) [inline unit/util/replace_symbol.test.cpp:74]: Shall these comments be removed?

File: unit/util/std_expr.test.cpp [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+  auto dividend = from_integer(10, t);
+  auto divisor = from_integer(5, t);
+  auto div = div_exprt(dividend, divisor);
+
+  SECTION("its divisor and dividend have the values assigned to them")
+  {
+    REQUIRE(div.op0() == dividend);
+    REQUIRE(div.op1() == divisor);
+  }
+  /* This will not work on current implementation (and we probably shouldn't waste time on this)
```
Comment (lucasccordeiro) [inline unit/util/std_expr.test.cpp:24]: Shall these comments be removed?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]

**Pull request merged**